pkg/dockerfile/fast_generator.go

```diff
@@ -0,0 +1,261 @@
+		return "", fmt.Errorf("Failed to write %s: %w", filename, err)
+	}
+	return path, nil
+}
+
+func (g *FastGenerator) generateMonobase(lines []string, tmpDir string) ([]string, error) {
+	lines = append(lines, []string{
+		"FROM r8.im/monobase:latest",
+	}...)
+
+	cogPath, err := g.copyCog(tmpDir)
+	if err != nil {
+		return nil, err
+	}
+
+	lines = append(lines, []string{
+		"ENV R8_COG_VERSION=\"file:///buildtmp/" + filepath.Base(cogPath) + "\"",
+	}...)
+
+	relativeTmpDir, err := filepath.Rel(g.Dir, tmpDir)
+	if err != nil {
+		return nil, err
+	}
+	skipCudaArg := "--skip-cuda"
+	if g.Config.Build.GPU {
+		skipCudaArg = ""
+		cudaVersion := g.Config.Build.CUDA
+		cudnnVersion := g.Config.Build.CuDNN
+		lines = append(lines, []string{
+			"ENV R8_CUDA_VERSION=" + cudaVersion,
+			"ENV R8_CUDNN_VERSION=" + cudnnVersion,
+			"ENV R8_CUDA_PREFIX=https://monobase.replicate.delivery/cuda",
+			"ENV R8_CUDNN_PREFIX=https://monobase.replicate.delivery/cudnn",
+		}...)
+	}
+
+	lines = append(lines, []string{
+		"ENV R8_PYTHON_VERSION=" + g.Config.Build.PythonVersion,
+	}...)
+
+	torchVersion, ok := g.Config.TorchVersion()
+	if ok {
+		lines = append(lines, []string{
+			"ENV R8_TORCH_VERSION=" + torchVersion,
+		}...)
+	}
+
+	return append(lines, []string{
+		"RUN --mount=type=bind,source=\"" + relativeTmpDir + "\",target=/buildtmp /opt/r8/monobase/build.sh " + skipCudaArg + " --mini",
+	}...), nil
+}
+
+func (g *FastGenerator) copyWeights(lines []string) ([]string, error) {
+	weights, err := FindWeights(g.Dir)
+	if err != nil {
+		return nil, err
+	}
+
+	if len(weights) == 0 {
+		return lines, nil
+	}
+
+	commands := []string{}
+	for sha256, file := range weights {
+		rel_path, err := filepath.Rel(g.Dir, file)
+		if err != nil {
+			return nil, err
+		}
+		commands = append(commands, "cp /src/"+rel_path+" "+filepath.Join(FUSE_RPC_WEIGHTS_PATH, sha256))
+	}
+
+	return append(lines, []string{
+		"RUN --mount=type=bind,ro,source=.,target=/src mkdir -p " + FUSE_RPC_WEIGHTS_PATH + " && " + strings.Join(commands, " && "),
+	}...), nil
+}
+
+func (g *FastGenerator) install(lines []string) ([]string, error) {
+	// Copy over source
+	commands := []string{
+		"mkdir /src && cp -r /srctmp /src && rm -rf /src/.cog",
+	}
+	mounts := []string{
+		"--mount=type=bind,ro,source=.,target=/srctmp",
+	}
+
+	// Install apt packages
+	packages := g.Config.Build.SystemPackages
+	if len(packages) > 0 {
+		mounts = append(mounts, "--mount=type=cache,target=/var/cache/apt,id=apt-cache")
+		aptCommand := "apt-get update && apt-get install -qqy "
+		aptCommand += strings.Join(packages, " ")
+		aptCommand += " && rm -rf /var/lib/apt/lists/*"
+		commands = append(commands, aptCommand)
+	}
+
+	// Install python packages
+	packages, err := g.pythonPackages()
+	if err != nil {
+		return nil, err
+	}
+	if len(packages) > 0 {
+		mounts = append(mounts, "--mount=type=cache,target=/root/.cache,id=pip-cache")
+		pipCommand := "uv pip install "
```

Comment:
This could install package already in monobase right? Like if user include `python_packages: torch==...` instead of having it under `torch: ...`.

https://github.com/replicate/monobase/blob/main/mini.Dockerfile#L24
Monobase build script can take a user `--requirements` arg which exclude pkgs already in monobase's corresponding `requirements.txt`. This ensures user venv is mutually exclusive from monobase venv.

BTW where/how is user venv created?

Reply:
Interestingly there isn't currently one, but I wonder whether thats ok? I think the layer should reflect the changes to the main venv which might be enough for the fuse config? lmk if thats a good assumption

Reply:
We need to create a new user venv. The monobase venv is in `/srv/r8/monobase/venv/...` where `/srv/r8/monobase` is a read-only volume mount. So we can't untar user files on top.

The dedup part (removing pkgs already in monobase) is so that when we can layer `PYTHONPATH=<monobase>:<user>:<cog>` because:
* To ensure user always depend on pkg version in monobase, not random ones pulled in by pip
* We can bump pkgs in monobase for security fixes, and be sure that it's picked up